### DISCUSSION TITLE
RFC on-chain storage

### DISCUSSION THREAD

**Discussion by gregLibert** [massalabs/massa#2668] [Ideas]
# On-chain storage
Normalization of binary storage in the Massa's datastore.

### Context
Storing data in the ledger for a given address is one of the functionalities proposed by the Massa blockchain.

Currently, the Massa's API exposes a key/value set by address where:
- keys are 256 bits hash value,
- values are byte arrays.

### Motivation
Sharing the same convention to interact with stored data will make content rendering easier and reduce heterogeneity.

### Scope
This RFC will only describe the storage envelope, not the content itself.

### Key
The following keys are reserved values:

| key | hash | Content |
| --- | ---- | ------- |
| Domain name resolution | 4226d14bcb58e71c97b4ef6415c6b73471dfe40f42d42f13f72b4b3fedca4473 | Used to resolve a domain name to a Massa address. |
| Proprietary | b904752a04b46e2865b8ea090d74154509bad4f2c84dde3326dc916efc2c58ad | Used to store data in a proprietary format. |
| Smart contract | f5674ec173cb2c99babcbf807a717dab78abd3e87f340d42bfd6bb73722c43ac | Used to store a smart contract. See https://github.com/massalabs/massa/issues/2642 for details. |
| Website                | b5a229ac8becc6035511f432ca6018f581f0627233eada6ae8e12b505d44af7f | Used to store a web site on-chain. |

Note: this list will be extended as the use cases evolve.

### Value
Each value is structured as follows:
- Version
This field (one byte) specifies which version of the storage convention is used to store the data. At the moment, 0x00 shall be used.
- Content
Binary data (byte array). For more information, please refer to the RFC describing the format in question.

Note: When possible the [TLV protocol](https://en.wikipedia.org/wiki/Type%E2%80%93length%E2%80%93value), with the Type field encoded in 2 bytes and the Length one encoded in 3 bytes, should be used.

**Comment by sebastien-forestier** [comment]
I don't understand what "Domain name resolution" is used for

**Reply by gregLibert** [reply]
Instead of using DNS, which is a loaded term IMHO, I used _Domaine name resolution_ to refer to a specific step of DNS process : [address resolution](https://en.wikipedia.org/wiki/Domain_Name_System#Address_resolution_mechanism).

The use cases that I have I mind are 
- using friendly name in URL (e.g. massa://gol)
- using friendly name in credit transfer (e.g. I want to send massa to sebastien-forestier)

Without this kind of functionality, addresses must be used...

**Reply by damip** [reply]
the problem is that the "lookup" direction is not right: we need an index that matches a domain name to an address. In this proposal, it's the address that's matched to the domain name

**Reply by gregLibert** [reply]
OK, I just understood the source of the misunderstanding : this RFC is about describing storage envelope, not the content itself.

_DNR_ and _website_ are ones of many possible contents. I could have proposed _media_ (video, image, sound), _documents_ ... and reserve a specific key to identify each one.

So to answer your question @sebastien-forestier , _DNR_ details should be defined in another RFC. This is also true for _website_ by the way.

**Reply by sebastien-forestier** [reply]
Like this it seems like everyone can choose the same domain name

**Comment by damip** [comment]
One thing missing is the capability to store bytecode. See: https://github.com/massalabs/massa/issues/2642

**Reply by gregLibert** [reply]
Can you clarify bytecode ? Is it a smart contract executable by the node using CallSC or some random code that can be downloaded and executed in a browser ?

Should we differentiate programming language at this level (one key for WebAssembly, another for js...) or simply use executable at this level and use tags to differentiate them ? I do have a preference for option 2.

**Reply by gregLibert** [reply]
Description updated with the _Smart contract_ key.

**Comment by gregLibert** [comment]
I try to clarify the scope of this RFC by adding:
- a scope
- a direct reference to additional RFC in _Content_ field

@damip is it more intelligible?

**Reply by damip** [reply]
yes, it's clear to me now :)

So for info, the table in the "key" subsection is given only as a list of random examples of reserved keys, it has no face value
